## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [3]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [5]:
# What three item categories had the most entries?
three_most_entries = df['category'].value_counts()
three_most_entries.head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [11]:
# For the category with the most entries, which subcategory had the most entries?
consumables_df = df[df['category'] == "consumables"]
subcategory_counts = consumables_df['subcategory'].value_counts()
subcategory_highest = subcategory_counts.idxmax()
highest_count = subcategory_counts.max()
print(f"The subcategory with the most entries in 'consumables' is '{subcategory_highest}' with {highest_count} entries.")   

The subcategory with the most entries in 'consumables' is 'bathroom supplies' with 6424 entries.


In [13]:
# Which five clients had the most entries in the data?
five_highest_clients = df['client_id'].value_counts()
five_highest_clients.head()

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [21]:
# Store the client ids of those top 5 clients in a list.
five_highest_clients = df['client_id'].value_counts().head(5)
top_five_client_ids = five_highest_clients.index.tolist()


The top 5 client IDs are: [33615, 66037, 46820, 38378, 24741]


In [22]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_id =df['client_id'].value_counts().idxmax()
top_client_orders = df[df['client_id'] == top_client_id]
total_units_top_client = top_client_orders['qty'].sum()
print("Total units of the client with the most entries")
print(total_units_top_client)

Total units of the client with the most entries
64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [24]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  line_subtotal  
0     1096.80     762.71         7.50  105            1      115164.00  
1       24.95      15.09         1.49   21            0         523.95  
2       13.52       7.86         1.68   39            6         527.28  
3       36.42      24.85         1.23   29            3        1056.18  
4      195.10     108.17        46.43   20            1        3902.00

In [27]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def calculate_shipping(row):
    total_weight = row['unit_weight'] * row['qty']
    if total_weight > 50:
        return 7 * total_weight
    else:
        return 10 * total_weight
    
df['shipping_price'] = df.apply(calculate_shipping, axis=1)

df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  line_subtotal  \
0     1096.80     762.71         7.50  105            1      115164.00   
1       24.95      15.09         1.49   21            0         523.95   
2       13.52       7.86         1.68   39            6         527.28   
3       36.42      24.85         1.23   29            3        1056.18   
4      195.10     108.17        46.43   20            1        3902.00   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20

In [28]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = (df['line_subtotal'] +df['shipping_price']) *1.0925
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number  line_subtotal  shipping_price  \
0     762.71         7.50  105            1      115164.00         5512.50   
1      15.09         1.49   21            0         523.95          312.90   
2       7.86         1.68   39            6         527.28          458.64   
3      24.85         1.23   29            3        1056.18          356.70   
4     108.17        46.43   20            1        3902.00         6500.20   

     total_price  
0  131839.076250  
1     914.258625  
2    1077.117600  
3    1543.571400  
4   11364.403500  

[5 rows x 21 columns]

In [29]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['total_cost'] = df['unit_cost'] *df['qty'] + df['shipping_price']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  line_subtotal  shipping_price  \
0         7.50  105            1      115164.00         5512.50   
1         1.49   21            0         523.95          312.90   
2         1.68   39            6         527.28          458.64   
3         1.23   29            3        1056.18          356.70   
4        46.43   20            1        3902.00         6500.20   

     total_price  total_cost  
0  131839.076250    85597.05  
1     914.258625      629.79  
2    1077.117600      765.18  
3    1543.571400     1077.35  
4   11364.403500     8663.60  

[5 rows x 22 columns]

In [31]:
# Create a column for the profit of each line using line cost and line price
df['profit'] = df['total_price'] - df['total_cost']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  line_subtotal  shipping_price    total_price  total_cost  \
0            1      115164.00         5512.50  131839.076250    85597.05   
1            0         523.95          312.90     914.258625      629.79   
2            6         527.28          458.64    1077.117600      765.18   
3            3        1056.18          356.70    1543.571400     1077.35   
4            1        3902.00         6500.20   11364.403500     8663.60   

         profit  
0  46242.026250  
1    284.468625  
2    311.937600  
3    466.221400  
4   2700.803500  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [33]:
# Check your work using the totals above
order_total_first = df[df['order_id'] == 2742071]['total_price'].sum()
order_total_second = df[df['order_id'] == 2173913]['total_price'].sum()
order_total_third = df[df['order_id'] == 6128929]['total_price'].sum()

print(f"The total price for order ID 2742071 is: ${order_total_first:.2f}")
print(f"The total price for order ID 2173913 is: ${order_total_second:.2f}")
print(f"The total price for order ID 6128929 is: ${order_total_third:.2f}")

The total price for order ID 2742071 is: $152811.90
The total price for order ID 2173913 is: $162388.72
The total price for order ID 6128929 is: $923441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [36]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_clients_df = df[df['client_id'].isin(top_five_client_ids)]
client_spending = top_clients_df.groupby('client_id')['total_price'].sum().reset_index()
client_spending.columns = ['client_id', 'total_spent']
print(client_spending)


   client_id   total_spent
0      24741  8.226889e+07
1      33615  8.377309e+06
2      38378  1.290655e+07
3      46820  9.743794e+06
4      66037  1.025951e+07


In [37]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
summary_df = top_clients_df.groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('shipping_price', 'sum'),
    total_revenue=('total_price', 'sum'),
    total_profit=('profit', 'sum') 
).reset_index()

summary_df.head()

client_id  total_units_purchased  total_shipping_price  total_revenue  \
0      24741                 239862            5126448.37   8.226889e+07   
1      33615                  64313            1828984.89   8.377309e+06   
2      38378                  73667            3429455.40   1.290655e+07   
3      46820                  75768            1601448.84   9.743794e+06   
4      66037                  43018            1395151.85   1.025951e+07   

   total_profit  
0  3.657999e+07  
1  2.201995e+06  
2  3.271830e+06  
3  2.736603e+06  
4  3.255032e+06

In [48]:
# Format the data and rename the columns to names suitable for presentation.
summary_df.columns = ['Client ID', 'Total Units Purchased', 'Total Shipping Cost ($)', 'Total Revenue ($)', 'Total Profit ($)']
# Define the money columns.
summary_df['Total Shipping Cost ($)'] = summary_df['Total Shipping Cost ($)']
summary_df['Total Revenue ($)'] = summary_df['Total Revenue ($)']
summary_df['Total Profit ($)'] = summary_df['Total Profit ($)']


# Define a function that converts a dollar amount to millions.
def currency_format_millions(dollar_amount):
    if dollar_amount is None:
        return None
    millions = dollar_amount / 1000000
    return millions

# Apply the currency_format_millions function to only the money columns. 
money_columns = ['Total Shipping Cost ($)', 'Total Revenue ($)', 'Total Profit ($)']
for column in money_columns:
    summary_df[column] = summary_df[column].apply(currency_format_millions)








TypeError: unsupported operand type(s) for //: 'str' and 'int'

In [ ]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
